# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/vr/ipc/VRManagerChild.cpp
# Commit: 51b0f3fef43a
# Full Hash: 51b0f3fef43ade362807e69e91d6786390e67015
# Author: sotaro <sotaro.ikeda.g@gmail.com>
# Date: 2020-02-15 09:56:17
# Description:
#   Bug 1614570 - Make VRManagerChild life time handling more like CompositorManagerChild r=aosmond
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D62710
# ==============================================================================

diff -r 10d15000963d -r 51b0f3fef43a gfx/vr/ipc/VRManagerChild.cpp
--- a/gfx/vr/ipc/VRManagerChild.cpp	Fri Feb 14 23:15:28 2020 +0000
+++ b/gfx/vr/ipc/VRManagerChild.cpp	Fri Feb 14 15:10:40 2020 +0000
@@ -43,6 +43,7 @@
   MOZ_ASSERT(NS_IsMainThread());
 
   mStartTimeStamp = TimeStamp::Now();
+  AddRef();
 }
 
 VRManagerChild::~VRManagerChild() { MOZ_ASSERT(NS_IsMainThread()); }
@@ -90,7 +91,6 @@
 /* static */
 bool VRManagerChild::InitForContent(Endpoint<PVRManagerChild>&& aEndpoint) {
   MOZ_ASSERT(NS_IsMainThread());
-  MOZ_ASSERT(!sVRManagerChildSingleton);
 
   RefPtr<VRManagerChild> child(new VRManagerChild());
   if (!aEndpoint.Bind(child)) {
@@ -100,15 +100,6 @@
   return true;
 }
 
-/* static */
-bool VRManagerChild::ReinitForContent(Endpoint<PVRManagerChild>&& aEndpoint) {
-  MOZ_ASSERT(NS_IsMainThread());
-
-  ShutDown();
-
-  return InitForContent(std::move(aEndpoint));
-}
-
 /*static*/
 void VRManagerChild::InitSameProcess() {
   MOZ_ASSERT(NS_IsMainThread());
@@ -135,21 +126,16 @@
 /*static*/
 void VRManagerChild::ShutDown() {
   MOZ_ASSERT(NS_IsMainThread());
-  if (sVRManagerChildSingleton) {
-    sVRManagerChildSingleton->Destroy();
-    SpinEventLoopUntil([&]() { return !sVRManagerChildSingleton; });
+  if (!sVRManagerChildSingleton) {
+    return;
   }
+  sVRManagerChildSingleton->Close();
+  sVRManagerChildSingleton = nullptr;
 }
 
-void VRManagerChild::Destroy() {
-  // Keep ourselves alive until everything has been shut down
-  RefPtr<VRManagerChild> selfRef = this;
-  MessageLoop::current()->PostTask(NewRunnableMethod(
-      "VRManagerChild::AfterDestroy", selfRef, &VRManagerChild::AfterDestroy));
-}
+void VRManagerChild::ActorDealloc() { Release(); }
 
-void VRManagerChild::AfterDestroy() {
-  Close();
+void VRManagerChild::ActorDestroy(ActorDestroyReason aReason) {
   if (sVRManagerChildSingleton == this) {
     sVRManagerChildSingleton = nullptr;
   }